### PR TITLE
refactor/mutable URLRequest

### PR DESCRIPTION
**Pull request by kpcwatson**
Get rid of `NSMutableURLRequest`

### CODE REVIEW

File: Horatio/Horatio/Classes/JSON/JSONServiceRequest.swift [hunk 1]
```diff
@@ -72,10 +73,12 @@ open class JSONBodyParametersServiceRequestDecorator: ServiceRequestDecorator {
...
     
-    open func compose(_ urlRequest: NSMutableURLRequest) {
+    open func compose(_ urlRequest: URLRequest) -> URLRequest {
+        var request = urlRequest
         do {
             let data = try JSONSerialization.data(withJSONObject: parameters, options: .prettyPrinted)
-            urlRequest.httpBody = data
+            request.httpBody = data
         } catch { }
+        return request
```
Comment (ktatroe) [inline Horatio/Horatio/Classes/JSON/JSONServiceRequest.swift]: `return` statements should always be their own paragraph; line break before here.

File: Horatio/Horatio/Classes/Services/ServiceRequestDecorator.swift [hunk 2]
```diff
@@ -27,8 +27,10 @@ open class AcceptGZIPHeadersServiceRequestDecorator: ServiceRequestDecorator {
 
     // MARK: <ServiceRequestDecorator>
 
-    open func compose(_ urlRequest: NSMutableURLRequest) {
-        urlRequest.setValue("gzip", forHTTPHeaderField: "Accept-Encoding")
+    open func compose(_ urlRequest: URLRequest) -> URLRequest {
+        var request = urlRequest
+        request.setValue("gzip", forHTTPHeaderField: "Accept-Encoding")
+        return request
```
Comment (ktatroe) [inline Horatio/Horatio/Classes/Services/ServiceRequestDecorator.swift]: Paragraph here.

File: Horatio/Horatio/Classes/Services/ServiceRequestDecorator.swift [hunk 3]
```diff
@@ -133,14 +137,17 @@ open class HTTPAuthServiceRequestDecorator: ServiceRequestDecorator {
     }
     
     
-    open func compose(_ urlRequest: NSMutableURLRequest) {
-        let credentials = "\(username):\(password)"
+    open func compose(_ urlRequest: URLRequest) -> URLRequest {
+        var request = urlRequest
         
+        let credentials = "\(username):\(password)"
```
Comment (ktatroe) [inline Horatio/Horatio/Classes/Services/ServiceRequestDecorator.swift]: Line-break here.

**Comment by kpcwatson** [comment]
`URLRequest` is a value type and isn't being retained on return. This is using a little more stack memory, but it should be negligible. When `URLRequest` is passed as a parameter or returned, it is a copy (sort've... because copy-on-write).

I only made this change to be "more swifty" and to prefer value over reference semantics. It uses more memory, but it's "safer" to use a value type (in general, though I doubt `ServiceRequestDecorator`s are being shared across threads... and I don't really see much issue even if they are). The memory usage should be pretty insignificant, especially since it's on the stack and transient.

That being said, `URLRequest` bridges to `NSURLRequest` and `NSMutableURLRequest` (and it looks like `URLRequest` keeps a copy of `NSMutableRequest`). Because of this, I can't think of any benefit to using `URLRequest` other than that it follows common swift paradigms. I just got in the habit of replacing `let|var blah = NSMutableWhatever` with `var blah = Whatever`... I thought it was because of a swiftlint rule, but I'm not seeing it. I'm fine with closing this PR.

**Comment by ktatroe** [comment]
Nah, I agree. The NSMutable* semantics are holdover from ObjC, and I prefer switching over as we can to more Swiftly things.

**Pull request merged**